### DISCUSSION TITLE
fp-ts and ChatGPT

### DISCUSSION THREAD

**Discussion by gcanti** [gcanti/fp-ts#1828] [General]
I opened this discussion to share ideas, prompts, and tips on how to use ChatGPT as a aid for development and understanding of the fp-ts APIs.

**Comment by bigpopakap** [comment]
The following example is about dev tooling that would help writing code using `fp-ts`, not how `fp-ts` could be used in GPT applications. Hopefully this fits with what you're looking for!

I often find myself writing a `pipe()` from both sides, and getting stuck in the middle when I have to do some fancy flattening/sequence stuff to get the types to work

Vaguely, I'll have a situation like this:
```ts
pipe(
  // The input is of type A, and the way this is used in the rest of my code, it's fixed
  // do some computations

  // I often write this to inspect what type I have
  // Here, say I have something like Either<string, ((_: A) => Task<ReadonlyNonemptyArray<A>>>
  a => a,

  // Now I just have to do some various sequence(), flatten(), toArray(), mapLeft()s to convert the current value to the known output type

  // The way this function is used, the output has to be (_: A) => TaskEither<[A, string], ReadonlyArray<A>>
)
```
The exact types are contrived, but I find myself in this situation often.

When I find myself here, I dream of using an AI that understands the types, and some basic functional programming patterns to autocomplete this for me

**Comment by bigpopakap** [comment]
Another idea: can GPT write the documentation site?

The serie of guides you've written on dev.to are great because they provide motivation and real life examples.

The documentation site is great once you already have that background, and are just looking up the exact types, or the exact name of a function.

If GPT could write the documentation site, it could achieve both in one place. It can provide an example usage of the `chain` function, explain how it suffers from `map`, etc.

This would be _most_ beneficial when Typescript forces the type-level stuff to get complicated, like with the `KindX<...>` types. Those are so abstract, it's hard to mentally map them to a concrete use case when reading the docs (ex. https://gcanti.github.io/fp-ts/modules/EitherT.ts.html#altvalidation)
